velox/docs/develop/aggregate-functions.rst

```diff
@@ -109,24 +109,323 @@ accordingly.  Data is stored in the following order:
 .. image:: images/aggregation-layout.png
   :width: 600
 
-Aggregate class
----------------
-
-To add an aggregate function,
+To add an aggregate function, there are two options: implementing it as a
+simple function or as a vector function. The simple-function interface allows
+the author to write methods that process input data one row at a time and not
+handle input vector encodings themselves. However, the simple-function
+interface currently has certain limitations, such as not allowing for advanced
+performance optimization on constant inputs. Aggregation functions that
+require such functionalities can be implemented through the vector-function
+interface. With the vector-function interface, the author writes methods that
+process one vector at a time and handles input vector encodings by themselves.
 
 * Prepare:
     * Figure out what are the input, intermediate and final types.
     * Figure out what are partial and final calculations.
     * Design the accumulator. Make sure the same accumulator can accept both raw
       inputs and intermediate results.
-    * Create a new class that extends velox::exec::Aggregate base class
+    * If implementing a simple function, create a class for the function according
+      to instructions below; If implementing a vector function,
+      create a new class that extends velox::exec::Aggregate base class
       (see velox/exec/Aggregate.h) and implement virtual methods.
 * Register the new function using exec::registerAggregateFunction(...).
 * Add tests.
 * Write documentation.
 
+Simple Function Interface
+-------------------------
+
+This section describes the main concepts and the simple interface of
+aggregation functions. Examples of aggregation functions implemented through
+the simple-function interface can be found at velox/exec/tests/SimpleAverageAggregate.cpp
+and velox/exec/tests/SimpleArrayAggAggregate.cpp.
+
+A simple aggregation function is implemented as a class as the following.
+
+.. code-block:: c++
+
+  // array_agg(T) -> array(T) -> array(T)
```

Comment:
Have another question : If there is an aggregate function with 2 different signatures then what are the recommendations ? How do we accomodate them in the same aggregate class or they need to be written as a second aggregate function ?

Reply:
Hi @aditi-pandit, If the two signatures have the same default-null-behavior flag (i.e., both true or both false), I think it's OK to implement them in the same class. (I haven't tried this yet but I think it's possible.) They need to have the same default-null-behavior flag because this affect the signatures of methods in the aggregate class. Whether implementing them in the same class or as separate classes is preferred depends on which implementation is cleaner in my opinion.

Reply:
If the number of arguments are the same between the two groups of function signatures, then the aggregate class can be templated on the input types and in the toIntermediate() (or other method) function, the code can check the input types using std::is_same_v. If the number of arguments are different, the aggregate class can be templated on a parameter pack. It would be more complex but should still be possible to implement the methods by checking the arity of the pack. But it probably doesn't worth the complexity. 

So I agree, having separate classes may be cleaner and more readable.